File: src/loop/start.ts
import crypto from "node:crypto";
import fs from "node:fs";

import type { ResolvedConfig } from "../config.js";
import { GymratError } from "../errors.js";
import { archivedSessionPath, sessionJsonlPath } from "../session/paths.js";
import type { SessionRecord } from "../session/records.js";
import { appendRecord, foldSession, readRecords, type SessionState } from "../session/store.js";
import { createWorkspace, detectWorkspace, recreateWorkspace } from "../session/workspace.js";
import { resolveTarget } from "../targets.js";

/** Ref the baseline is pinned to when the caller names none. */
const DEFAULT_BASELINE_REF = "HEAD";

/** Hex digits distinguishing two sessions started within the same second. */
const SESSION_ID_ENTROPY_BYTES = 2;

/** A session ready to iterate in, together with everything its log already holds. */
export interface StartResult {
  /** The header the session's log opens with, newly written or read back. */
  session: SessionRecord;
  /** What the log adds up to, so a resumed session can report its history. */
  state: SessionState;
  /** Whether the session was already on disk. */
  resumed: boolean;
  /** The id of the finalized session whose log was moved aside to make room for this one. */
  archived?: string;
  /** The path the finalized session's log was moved to. */
  archivedPath?: string;
}

/**
 * Create the repository's optimization session, or resume the one it already has.
 *
 * Resuming is what makes a second `start` safe: an existing log is never appended
 * to, and only a worktree that went missing is put back. A finalized session is
 * the one exception — it is closed, so its log is moved aside and a fresh session
 * takes its place rather than the agent being refused until it deletes something.
 * Holding the repository lock across the call is the caller's job — this touches
 * `.gymrat/` and the repository's branches, so two concurrent runs must not reach it.
 *
 * @throws GymratError when `ref` names no commit, when the log is corrupt, or
 *   when git refuses to create or recreate the workspace.
 */
export function startSession(
  root: string,
  ref: string | undefined,
  config: ResolvedConfig,
): StartResult {
  const jsonlPath = sessionJsonlPath(root);
  const records = readRecords(jsonlPath);
  const state = foldSession(records);
  const { session } = state;
  const baselineRef = ref ?? DEFAULT_BASELINE_REF;

  if (session === undefined) {
    return createSession(root, jsonlPath, baselineRef, config);
  }

  if (state.finalized !== undefined) {
    // Renamed rather than copied: the new log must not exist until the closed one
    // is out of the way, or a start interrupted mid-archive would leave two
    // sessions claiming the same file. A start that then fails renames it back,
    // so the closed session stays where `status` reads it — safe because the new
    // header lands last, leaving nothing at `jsonlPath` to collide with.
    const archivedPath = archivedSessionPath(root, session.sessionId);
    fs.renameSync(jsonlPath, archivedPath);
    try {
      return {
        ...createSession(root, jsonlPath, baselineRef, config),
        archived: session.sessionId,
        archivedPath,
      };
    } catch (error) {
      restoreArchivedLog(archivedPath, jsonlPath);
      throw error;
    }
  }

  if (!detectWorkspace(root)) {
    // Every keep moves the baseline onto the commit it made, so a baseline worktree
    // put back at the header's pinned SHA would have the next iteration measure the
    // whole session's diff instead of the edit in front of it.
    recreateWorkspace(root, session.branch, state.lastKeptCommit ?? session.baseline.sha);
  }
  return { session, state, resumed: true };
}

/**
 * Move a closed session's log back from the archive after a start that failed.
 *
 * Best-effort: the caller is rethrowing the failure that broke the start, and a
 * rename that cannot run must not speak in its place — the closed session's
 * records are still on disk under its own id either way.
 */
function restoreArchivedLog(archivedPath: string, jsonlPath: string): void {
  try {
    fs.renameSync(archivedPath, jsonlPath);
  } catch {
    // Swallowed by contract — see above.
  }
}

/**
 * Pin the baseline, build the workspace, and write the header that opens the log.
 *
 * The header lands last: a session the log claims exists but whose branch git
 * never created would send every later command looking for a workspace that
 * is not there.
 */
function createSession(
  root: string,
  jsonlPath: string,
  ref: string,
  config: ResolvedConfig,
): StartResult {
  const sha = resolveBaselineSha(ref, root);
  const now = new Date();
  const sessionId = newSessionId(now);
  const workspace = createWorkspace(root, sessionId, { ref, sha });

  const session: SessionRecord = {
    type: "session",
    schemaVersion: 1,
    sessionId,
    createdAt: now.toISOString(),
    baseline: workspace.baseline,
    branch: workspace.branch,
    worktrees: workspace.worktrees,
    config: snapshotConfig(config),
  };
  appendRecord(jsonlPath, session);

  return { session, state: foldSession([session]), resumed: false };
}

/**
 * The commit `ref` names, peeled through the same resolution `compare` uses.
 *
 * A baseline is always a ref: the session's own worktrees are where benchmarks
 * run, so a directory — which {@link resolveTarget} otherwise accepts, and
 * prefers over a ref of the same name — has nothing to pin a branch at.
 *
 * @throws GymratError when `ref` resolves to a directory or to no commit at all.
 */
function resolveBaselineSha(ref: string, root: string): string {
  const target = resolveTarget(ref, root);
  if (target.kind !== "ref") {
    throw new GymratError(
      `Cannot start a session at '${ref}': it names a directory, not a git ref`,
      "Pass a branch, tag, or commit the session's baseline is pinned to.",
    );
  }
  return target.resolvedSha;
}

/**
 * `<YYYYMMDD-HHmmss>-<4 hex>` in UTC.
 *
 * The timestamp sorts sessions the way they were started and reads back as a
 * date; the random suffix keeps two sessions started in the same second — and
 * therefore their branches — apart.
 */
function newSessionId(now: Date): string {
  const iso = now.toISOString();
  return `${iso.slice(0, 10).replaceAll("-", "")}-${iso.slice(11, 19).replaceAll(":", "")}-${crypto.randomBytes(SESSION_ID_ENTROPY_BYTES).toString("hex")}`;
}

/**
 * The settings the header records as provenance.
 *
 * Only the keys the session schema declares survive, and an absent optional key
 * stays absent rather than becoming an explicit `undefined`. Every command
 * re-reads `gymrat.json` for the settings it acts on, so this snapshot answers
 * "what was this session started with", never "what runs next".
 */
function snapshotConfig(config: ResolvedConfig): SessionRecord["config"] {
  return {
    bench: config.bench,
    ...(config.prepare === undefined ? {} : { prepare: config.prepare }),
    adapter: config.adapter,
    samples: config.samples,
    timeoutSeconds: config.timeoutSeconds,
    primary: config.primary,
    ...(config.filter === undefined ? {} : { filter: config.filter }),
    ...(config.hooks === undefined ? {} : { hooks: config.hooks }),
  };
}

File: src/gymrat/loop/start.py
"""Create a repository's optimization session, or resume the one it already has.

Resuming is what makes a second ``start`` safe: an existing log is never appended
to, and only a worktree that went missing is put back. A finalized session is the
one exception — it is closed, so its log is moved aside and a fresh session takes
its place rather than the agent being refused until it deletes something. Holding
the repository lock across the call is the caller's job — this touches ``.gymrat/``
and the repository's branches, so two concurrent runs must not reach it.
"""

import contextlib
import secrets
from dataclasses import dataclass
from datetime import UTC, datetime
from pathlib import Path

from gymrat.config import ResolvedConfig
from gymrat.errors import GymratError
from gymrat.session.clock import format_iso
from gymrat.session.paths import archived_session_path, session_jsonl_path
from gymrat.session.records import SCHEMA_VERSION, SessionConfig, SessionHooks, SessionRecord
from gymrat.session.store import SessionState, append_record, fold_session, read_records
from gymrat.session.workspace import BaselineRef, create_workspace, recreate_workspace
from gymrat.targets import RefTarget, resolve_target

# Ref the baseline is pinned to when the caller names none.
DEFAULT_BASELINE_REF = "HEAD"

# Bytes of entropy distinguishing two sessions started within the same second.
SESSION_ID_ENTROPY_BYTES = 2


@dataclass(frozen=True, slots=True)
class StartResult:
    """A session ready to iterate in, together with everything its log already holds."""

    #: The header the session's log opens with, newly written or read back.
    session: SessionRecord
    #: What the log adds up to, so a resumed session can report its history.
    state: SessionState
    #: Whether the session was already on disk.
    resumed: bool
    #: The id of the finalized session whose log was moved aside for this one.
    archived: str | None = None
    #: The path the finalized session's log was moved to.
    archived_path: str | None = None


def start_session(root: str, ref: str | None, config: ResolvedConfig) -> StartResult:
    """Create the repository's optimization session, or resume the one it already has.

    Args:
        root: The repository the session lives in.
        ref: The ref the baseline is pinned to, or ``None`` to pin at ``HEAD``.
        config: The resolved run settings the header snapshots as provenance.

    Raises:
        GymratError: When ``ref`` names no commit or a directory, when the log is
            corrupt, or when git refuses to create or recreate the workspace.
    """
    jsonl_path = session_jsonl_path(root)
    state = fold_session(read_records(jsonl_path))
    session = state.session
    baseline_ref = ref if ref is not None else DEFAULT_BASELINE_REF

    if session is None:
        return _create_session(root, jsonl_path, baseline_ref, config)

    if state.finalized is not None:
        # Renamed rather than copied: the new log must not exist until the closed
        # one is out of the way, or a start interrupted mid-archive would leave two
        # sessions claiming the same file. A start that then fails renames it back,
        # so the closed session stays where ``status`` reads it — safe because the
        # new header lands last, leaving nothing at ``jsonl_path`` to collide with.
        archived_path = archived_session_path(root, session.session_id)
        Path(jsonl_path).rename(archived_path)
        try:
            created = _create_session(root, jsonl_path, baseline_ref, config)
        except BaseException:
            _restore_archived_log(archived_path, jsonl_path)
            raise
        return StartResult(
            session=created.session,
            state=created.state,
            resumed=False,
            archived=session.session_id,
            archived_path=archived_path,
        )

    # Every keep moves the baseline onto the commit it made, so a baseline worktree
    # put back at the header's pinned SHA would have the next iteration measure the
    # whole session's diff instead of the edit in front of it. recreate_workspace
    # already no-ops when both worktrees stand, so no on-disk check is needed here.
    recreate_workspace(
        root,
        session.branch,
        state.last_kept_commit if state.last_kept_commit is not None else session.baseline.sha,
    )
    return StartResult(session=session, state=state, resumed=True)


def _restore_archived_log(archived_path: str, jsonl_path: str) -> None:
    """Move a closed session's log back from the archive after a start that failed.

    Best-effort: the caller is re-raising the failure that broke the start, and a
    rename that cannot run must not speak in its place — the closed session's
    records are still on disk under its own id either way.
    """
    # Swallowed by contract — see above.
    with contextlib.suppress(OSError):
        Path(archived_path).rename(jsonl_path)


def _create_session(root: str, jsonl_path: str, ref: str, config: ResolvedConfig) -> StartResult:
    """Pin the baseline, build the workspace, and write the header that opens the log.

    The header lands last: a session the log claims exists but whose branch git
    never created would send every later command looking for a workspace that is
    not there.
    """
    sha = _resolve_baseline_sha(ref, root)
    now = datetime.now(UTC)
    session_id = _new_session_id(now)
    workspace = create_workspace(root, session_id, BaselineRef(ref=ref, sha=sha))

    session = SessionRecord(
        type="session",
        schema_version=SCHEMA_VERSION,
        session_id=session_id,
        created_at=format_iso(now),
        baseline=workspace.baseline,
        branch=workspace.branch,
        worktrees=workspace.worktrees,
        config=_snapshot_config(config),
    )
    append_record(jsonl_path, session)

    return StartResult(session=session, state=fold_session([session]), resumed=False)


def _resolve_baseline_sha(ref: str, root: str) -> str:
    """The commit ``ref`` names, peeled through the same resolution ``compare`` uses.

    A baseline is always a ref: the session's own worktrees are where benchmarks
    run, so a directory — which :func:`resolve_target` otherwise accepts, and
    prefers over a ref of the same name — has nothing to pin a branch at.

    Raises:
        GymratError: When ``ref`` resolves to a directory or to no commit at all.
    """
    target = resolve_target(ref, root)
    if not isinstance(target, RefTarget):
        raise _directory_baseline_error(ref)
    return target.resolved_sha


def _directory_baseline_error(ref: str) -> GymratError:
    """The failure a baseline that resolves to a directory is reported as."""
    return GymratError(
        f"Cannot start a session at '{ref}': it names a directory, not a git ref",
        hint="Pass a branch, tag, or commit the session's baseline is pinned to.",
    )


def _new_session_id(now: datetime) -> str:
    """``<YYYYMMDD-HHMMSS>-<4 hex>`` in UTC.

    The timestamp sorts sessions the way they were started and reads back as a
    date; the random suffix keeps two sessions started in the same second — and
    therefore their branches — apart.
    """
    suffix = secrets.token_hex(SESSION_ID_ENTROPY_BYTES)
    return f"{now.strftime('%Y%m%d-%H%M%S')}-{suffix}"


def _snapshot_config(config: ResolvedConfig) -> SessionConfig:
    """The settings the header records as provenance.

    Only the keys the session schema declares survive, and an optional that was
    never set stays absent rather than becoming an explicit null. Every command
    re-reads its config for the settings it acts on, so this snapshot answers
    "what was this session started with", never "what runs next".
    """
    hooks = config.hooks
    return SessionConfig(
        bench=config.bench,
        adapter=config.adapter,
        samples=config.samples,
        timeout_seconds=config.timeout_seconds,
        primary=config.primary,
        prepare=config.prepare,
        filter=config.filter,
        hooks=SessionHooks(before=hooks.before, after=hooks.after) if hooks is not None else None,
    )
